internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,606 @@
+				cancel()
+				// Wait for the cancel to go through
+				time.Sleep(time.Millisecond)
+			}
+
+			if tc.secondCall {
+				iaReq := &authd.IARequest{
+					SessionId:          tc.sessionID,
+					AuthenticationData: "some data",
+				}
+				iaResp, err := client.IsAuthorized(context.Background(), iaReq)
+				secondCall = fmt.Sprintf("SECOND CALL:\n\taccess: %s\n\tdata: %s\n\terr: %v\n",
+					iaResp.GetAccess(),
+					iaResp.GetData(),
+					err,
+				)
+			}
+			<-done
+			got := firstCall + secondCall
+			want := testutils.LoadWithUpdateFromGolden(t, got)
+			require.Equal(t, want, got, "IsAuthorized should return the expected combined data, but did not")
+		})
+	}
+}
+
+func TestEndSession(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		username  string
+		brokerID  string
+		sessionID string
+
+		wantErr bool
+	}{
+		"Successfully end session": {username: "success"},
+
+		"Error when sessionID is empty":    {sessionID: "empty", wantErr: true},
+		"Error when sessionID is invalid":  {sessionID: "invalid", wantErr: true},
+		"Error when broker does not exist": {sessionID: "no broker", brokerID: "does not exist", wantErr: true},
+		"Error when ending session":        {username: "ES_error", wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			brokerCfgDir := t.TempDir()
+			brokerCfg := testutils.StartBusBrokerMock(t, brokerCfgDir)
+
+			client := startClient(t, brokerCfgDir)
+
+			if tc.brokerID == "" {
+				tc.brokerID = getGeneratedID(t, client, filepath.Base(brokerCfg))
+			}
+
+			if tc.brokerID != "does not exist" {
+				if tc.username == "" {
+					tc.username = "does not matter"
+				}
+				sbResp, err := client.SelectBroker(context.Background(), &authd.SBRequest{
+					BrokerId: tc.brokerID,
+					Username: tc.username,
+				})
+				require.NoError(t, err, "Setup: failed to create session for tests")
+				if tc.sessionID == "" {
+					tc.sessionID = sbResp.GetSessionId()
+				}
+			}
+
+			if tc.sessionID == "empty" {
+				tc.sessionID = ""
+			}
+
+			esReq := &authd.ESRequest{
+				SessionId: tc.sessionID,
+			}
+			_, err := client.EndSession(context.Background(), esReq)
+			if tc.wantErr {
+				require.Error(t, err, "EndSession should return an error, but did not")
+				return
+			}
+			require.NoError(t, err, "EndSession should not return an error, but did")
+		})
+	}
+}
+
+// getGeneratedID returns the generated ID for the specified broker.
+func getGeneratedID(t *testing.T, client authd.PAMClient, brokerName string) (id string) {
+	t.Helper()
+
+	r, err := client.AvailableBrokers(context.Background(), &authd.Empty{})
+	require.NoError(t, err, "Setup: could not get available brokers for tests")
+
+	for _, b := range r.GetBrokersInfos() {
+		if b.GetName() != brokerName {
+			continue
+		}
+		id = b.Id
+	}
+
+	return id
+}
+
+func startClient(t *testing.T, brokerDir string) (client authd.PAMClient) {
+	t.Helper()
+
+	a := daemon.New()
```

Comment:
Ok, I understand now why you are exporting SetArgs. I don’t think we are at the right level of tests. Those tests should only create a local GRPC server and listen on it. We don’t want to import anything from cmd/ in internal/. Just create the simplest GRPC server and client that you can. That will simplify always the whole handling of brokerDir. But yeah, basically what you are doing here is almost some integration tests (save the code for it :p). We want package tests.